### ISSUE TITLE
[Bug]: Reply.CreateBufferredCopy Fails when using WebHttpBinding

### ISSUE THREAD

**Issue by thunderhole** [CoreWCF/CoreWCF#1434]
### Duplicate ?

- [X] I have searched issues/discussions and did not find other issues/discussions reporting this bug.

### Product version

1.5.2

### Describe expected behavior

Expectation is for Message(reply).CreateBufferredCopy to work as it does for SOAP messages using WSHttpBinding. 

### Describe actual behavior

An exception is generated when calling CreateBufferedCopy when using json with the WebHttpBinding.

This code is ported from a working WCF project where this message inspector logs correctly. The service works as expected if the TraceBehavior which calls the IDispatchMessageInspector is removed from the service. Additionally, the message inspector works for the SOAP binding and the original project.

### Which binding

WebHttp

### security

Transport

### Which .NET version

.NET 6

### Which os platform

Windows

### Code snippet used to reproduce the issue

```c#
public void BeforeSendReply(ref Message reply, object request)
{
    MessageContext? messageContext = request as MessageContext;
    messageContext?.stopwatch?.Stop();

    MessageBuffer buffer = reply.CreateBufferedCopy(Int32.MaxValue);
    Log(messageContext?.buffer, buffer, messageContext?.stopwatch?.ElapsedMilliseconds);
    reply = buffer.CreateMessage();
}

public void ApplyDispatchBehavior(ServiceEndpoint endpoint, EndpointDispatcher endpointDispatcher)
{
    MessageInspector inspector = new();
    endpointDispatcher.DispatchRuntime.MessageInspectors.Add(inspector);
}

private static Binding JSONBinding()
{
    WebHttpBinding baseBinding = new(WebHttpSecurityMode.Transport)
    {
        MaxReceivedMessageSize = 16777216
    };

    CustomBinding binding = new(baseBinding);
    binding.Elements.Find<WebMessageEncodingBindingElement>().ContentTypeMapper = new RawTypeMapper();
    return binding;
}

private class RawTypeMapper : WebContentTypeMapper
{
    public override WebContentFormat GetMessageFormatForContentType(string contentType)
    {
        return WebContentFormat.Raw;
    }
}

// All the endpoints in this service return a Stream
return new MemoryStream(Encoding.UTF8.GetBytes(result));
```

### Stacktrace if any

```shell
CoreWCF.QuotaExceededException
  HResult=0x80131500
  Message=The maximum message size quota for incoming messages (2147483647) has been exceeded. To increase the quota, use the MaxReceivedMessageSize property on the appropriate binding element.
  Source=CoreWCF.Primitives
  StackTrace:
   at CoreWCF.Runtime.BufferedOutputStream.WriteCore(Byte[] buffer, Int32 offset, Int32 size)
   at CoreWCF.Channels.XmlByteStreamWriter.WriteBase64(Byte[] buffer, Int32 index, Int32 count)
   at System.Xml.XmlDictionaryWriter.WriteValue(IStreamProvider value)
   at CoreWCF.Channels.ByteStreamMessage.InternalByteStreamMessage.StreamedBodyWriter.OnWriteBodyContents(XmlDictionaryWriter writer)
   at CoreWCF.Channels.ByteStreamMessage.InternalByteStreamMessage.StreamedBodyWriter.OnCreateBufferedCopy(Int32 maxBufferSize)
   at CoreWCF.Channels.BodyWriter.CreateBufferedCopy(Int32 maxBufferSize)
   at CoreWCF.Channels.ByteStreamMessage.InternalByteStreamMessage.OnCreateBufferedCopy(Int32 maxBufferSize)
   at MessageInspector.BeforeSendReply(Message& reply, Object request) in MessageInspector.cs:line 69
   at CoreWCF.Dispatcher.ImmutableDispatchRuntime.BeforeSendReplyCore(MessageRpc rpc, Exception& exception, Boolean& thereIsAnUnhandledException)
```

**Comment by mconnew** [comment]
I'm working on the unit tests and fix now.

**Comment by thunderhole** [comment]
Thanks,

I wasn't going to be able to get to this til this weekend.

**Comment by mconnew** [comment]
I think you can work around the problem by changing your contract to return a Message object, and then return a message like this:
```c#
var buffer = new ArraySegment<byte>(Encoding.UTF8.GetBytes(result));
return ByteStreamMessage.CreateMessage(buffer);
```
It uses a different underlying body writer (BufferedBodyWriter vs StreamedBodyWriter). This is actually more performant too, less memory allocation. When creating a message from a Stream, because it doesn't know if the stream is replayable. There is a property to ask if it's seekable, but for things like reading from a file, you can get different data if some other process is modifying it so it's still technically not replayable, and also can't trust the implementation if it's some random non-runtime implementation. Because of this, it writes the stream out to a buffered stream, then calls ToArray on that buffered stream and wraps it in another class (ByteStreamBufferedMessageData) which is what gets used if you create a message from an ArraySegment in the first place. When you have a Message backed by an ArraySegment (like what happens after the stream copy), creating a buffered copy is low allocation as it just copies the reference to the ByteStreamBufferedMessageData and there's no big allocation or copying of data. As you are always creating a buffered copy in your message inspector, I would recommend making this change anyway, and BufferedBodyWriter would work today as it doesn't have this bug.

**Comment by thunderhole** [comment]
Unfortunately the work around alters the output.

`<Binary>ew0KICAidmVyc2lvbiI6ICIwLjAuMSINCn0=</Binary>`

instead of the json

```
{
    "version": "0.0.1"
}
```

**Comment by mconnew** [comment]
How are you examining the output? If you are checking Message.ToString() (or looking at the Message object in a debugger), it's going to lie to you about what the output looks like. There should be no way for it to output the Binary tags as they don't actually exist, it's faked that way to make it look like an XML SOAP body. You can see that in the [code here](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Channels/XmlByteStreamReader.cs#L50) where when fetching the LocalName from the XmlReader, it returns `ByteStreamMessageUtility.StreamElementName`, which has the value `"Binary"`. When you use ToString on the Message class, it creates a regular text XmlWriter and copies the XmlReader to that XmlWriter, which is writing to a StringWriter I think. So you get the abstracted faked SOAP view of the message when you do that. The code for writing the message out in ByteStreamMessageEncoder [can be seen here](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Channels/ByteStreamMessageEncoder.cs#L131-L135). It creates an XmlByteStreamWriter and then calls Message.WriteMessageAsync. This will cause Message to write it out like a SOAP message which includes calling XmlWriter.WriteStartElement for the `Binary` node. If you look at the implementation of [XmlByteStreamWriter.WriteStartElement here](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Channels/XmlByteStreamWriter.cs#L185-L200) you can see that verifies and modifies the current state to make sure it's being used correctly, but doesn't actually write anything out. When it comes to writing out the buffer, as you saw from an XML standpoint byte arrays/streams have to be written as base64 as XML can't represent arbitrary bytes. If you were sending a regular SOAP message, this is exactly how a Stream or byte[] gets written to XML. But if you look at [the implementation of XmlByteStream.WriteBase64Async](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Channels/XmlByteStreamWriter.cs#L131-L137), it doesn't base64 encode the bytes and instead just writes them directly to the output stream.  

Basically I suspect you validated this the wrong way. If I'm mistaken, please let me know.

**Comment by thunderhole** [comment]
This was tested via Postman. The output I pasted was from the Postman
response.

**Comment by mconnew** [comment]
@thunderhole, can you test with v1.6.0 that was just released. I'm a little worried something else might be broken as the `<Binary>` xml element should never make it out to Postman.
